README.md
# react-optimized-images

[![npm version](https://badge.fury.io/js/react-optimized-images.svg)](https://badge.fury.io/js/react-optimized-images)

A package to optimize images for React! It generates and uses responsive webp images with fallback to other formats, resulting in great performance on modern browsers and still supporting old browsers.

NOTE: If you created your app with CRA and it's not ejected or you don't have webpack as your bundler, you can use [v1](https://www.npmjs.com/package/react-optimized-images/v/1.2.2).

## Table of contents

- [Table of contents](#table-of-contents)
- [Features](#features)
- [Requirements](#requirements)
- [Installation](#installation)
- [Usage](#usage)
  - [Webpack plugin](#webpack-plugin)
  - [Next.js plugin](#nextjs-plugin)
  - [Picture component](#picture-component)
- [License](#license)

## Features

- Convert images to webp during build time.
- Create smaller versions of images based on breakpoints to avoid unnecessary load of big images on smaller devices.
- Lazy load images that are not in the viewport.
- React Component to make use of all the features with ease.

NOTE: Currently only images that are located within the project will be optimized.

#### TODO

- Add image optimization for external images.
- Add image optimization for CSS background images.
- Add support to older Node versions.
- Add unit tests.
- Add custom configuration at component level.

## Requirements

- Node v14.0.0 or above installed.
- Webpack as your project bundler.

## Installation

```
npm install react-optimized-images --save
```

or

```
yarn add react-optimized-images
```

## Usage

### Webpack plugin

This package generates images at build time, so you need to add the webpack plugin in your `webpack.config.js`.

```javascript
const OptimizedImagesPlugin = require('react-optimized-images/plugin');

module.exports = {
  // ... webpack config
  plugins: [
    // ... other plugins
    new OptimizedImagesPlugin({ ...options }),
  ],
};
```

#### Options

| Name        | Type    | Default value                         | Description                                                                                                                                 |
| ----------- | ------- | ------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------- |
| minWidth    | number  | 200                                   | Mininum image width to create smaller versions.                                                                                             |
| breakpoints | array   | [Default value](#default-breakpoints) | Specifies the breakpoints to generate responsive images. You can add as many breakpoints as you want.                                       |
| enabled     | boolean | true                                  | If false, the images won't be generated and the component will render a regular image.                                                      |
| lazy        | boolean | false                                 | If true, all the images in the `Picture` component will be lazy loaded by default. It can be overriden by the `lazy` prop in the component. |

##### Default breakpoints

```javascript
[
  {
    maxWidth: 576, // Max screen width in px
    resizeTo: 50, // Percentage of original image to resize
    // Example: An image with 1000px would be resized to 500px in a 576px or smaller screen
  },
  {
    maxWidth: 992,
    resizeTo: 70,
  },
];
```

Complete custom configuration example:

```javascript
const OptimizedImagesPlugin = require('react-optimized-images/plugin');

module.exports = {
  // ... webpack config
  plugins: [
    // ... other plugins
    new OptimizedImagesPlugin({
      minWidth: 300,
      breakpoints: [
        {
          maxWidth: 576,
          resizeTo: 50,
        },
        {
          maxWidth: 992,
          resizeTo: 80,
        },
      ],
      enabled: process.env.NODE_ENV === 'production',
      lazy: false,
    }),
  ],
};
```

### Next.js plugin

This package also has a plugin for Next.js projects. In your `next.config.js` add:

```javascript
const withOptimizedImages = require('react-optimized-images/next');

module.exports = withOptimizedImages();
```

Example with custom configuration:

```javascript
const withOptimizedImages = require('react-optimized-images/next');

module.exports = withOptimizedImages({
  minWidth: 300,
  breakpoints: [
    {
      maxWidth: 576,
      resizeTo: 50,
    },
    {
      maxWidth: 992,
      resizeTo: 80,
    },
  ],
});
```

Or if you use `next-compose-plugins`:

```javascript
const withPlugins = require('next-compose-plugins');
const optimizedImages = require('react-optimized-images/next');

module.exports = withPlugins([
  // other plugins
  [
    optimizedImages,
    /* Custom configuration
    {
       minWidth: 300
    },
    */
  ],
]);
```

### Picture component

For an easier use of the generated images, you can use the `Picture` component, a wrapper for html `<picture>` tag. It looks for the optimized webp images and uses the default format as a fallback for older browsers, providing full support to all browsers.

#### Usage

```javascript
import React from 'react';
import { Picture } from 'react-optimized-images';

import CoffeeJpg from '../assets/coffee.jpg';

export default function Home() {
  return (
    <div>
      <Picture src={CoffeeJpg} />
      {/* ..with Next v11.. */}
      <Picture src={CoffeeJpg.src} />
      {/* ..or if it comes from public folder */}
      <Picture src="/coffee.jpg" />
    </div>
  );
}
```

The output will be like

```html
<picture>
  <source
    srcset="/coffee@0.5x.webp"
    media="(max-width: 576px)"
    type="image/webp"
  />
  <source
    srcset="/coffee@0.7x.webp"
    media="(max-width: 992px)"
    type="image/webp"
  />
  <source srcset="/coffee.webp" type="image/webp" />
  <source
    srcset="/coffee@0.5x.jpeg"
    media="(max-width: 576px)"
    type="image/jpeg"
  />
  <source
    srcset="/coffee@0.7x.jpeg"
    media="(max-width: 992px)"
    type="image/jpeg"
  />
  <source srcset="/coffee.jpg" type="image/jpeg" />
  <img src="/coffee.jpg" />
</picture>
```

#### Properties

It uses the same properties from HTML `<img>` tag. Along with some specific properties as shown below:

| Name | Type | Default value | Description |
| ---- | ---- | ------------- | ----------- |
| lazy | boolean | false | If true, the image will be loaded only when content is ready and the image is inside the viewport.
| preview | ReactNode | undefined | A custom preview to be shown during image load if `lazy` is true. If preview is not set and `height` and `width` props are defined, a blurry version of the original image will be shown.

## License

[MIT](https://github.com/MarceloDJunior/react-optimized-images/blob/main/LICENSE) © Marcelo Dornelles Junior


CHANGELOG.md
# [2.1.0](https://github.com/MarceloDJunior/react-optimized-images/compare/v2.0.1...v2.1.0) (2022-06-16)


### Features

* add custom preview during lazy load ([e5479eb](https://github.com/MarceloDJunior/react-optimized-images/commit/e5479eb064524e8e9452b35586d3f80fd7581227))
* add default lazy load option in webpack custom config ([0a19715](https://github.com/MarceloDJunior/react-optimized-images/commit/0a19715e56b5cd8d1fcf8ad61dcfbc335fa0afec))
* add lazy load to images ([60339c0](https://github.com/MarceloDJunior/react-optimized-images/commit/60339c07db54fcd2282bfa7105d81151d66a922a))
* add lazy option in picture props ([c092850](https://github.com/MarceloDJunior/react-optimized-images/commit/c092850caafb6e6651f7d920c9ea7bfef5b52200))
* generate blurred preview images ([0f8cac4](https://github.com/MarceloDJunior/react-optimized-images/commit/0f8cac4eb44f09904b903ccececdeb1ce83602c6))



## [2.0.1](https://github.com/MarceloDJunior/react-optimized-images/compare/v2.0.0...v2.0.1) (2022-06-07)


### Bug Fixes

* update npm version over unpublished one ([736f3cd](https://github.com/MarceloDJunior/react-optimized-images/commit/736f3cd41725bb0554340f79b467c4cfb9d0716c))



# [2.0.0](https://github.com/MarceloDJunior/react-optimized-images/compare/v1.2.2...v2.0.0) (2022-06-07)


### Bug Fixes

* add exports to enable jest mocks ([1a6dfe6](https://github.com/MarceloDJunior/react-optimized-images/commit/1a6dfe65dcd92a0cd29f2b66d33853470488b6e9))


### Features

* add enabled option ([479f340](https://github.com/MarceloDJunior/react-optimized-images/commit/479f3406b6757b5c9169ad326e7543aa548168cd))
* add images generation with custom configuration at build time ([cefb9d0](https://github.com/MarceloDJunior/react-optimized-images/commit/cefb9d0c4d806ff7984a3f9c6fe895828ed567b2))
* add Next.js plugin ([db657b7](https://github.com/MarceloDJunior/react-optimized-images/commit/db657b75695afda24a64db1707cbef61c6f67b5b))


### BREAKING CHANGES

* `convert-images` script is no longer supported. Please add `react-optimized-images/plugin` in your webpack config to generate optimized images. Instructions are available in the README.



## [1.2.2](https://github.com/MarceloDJunior/react-optimized-images/compare/v1.2.1...v1.2.2) (2022-05-29)


### Bug Fixes

* add className to picture component to avoid wrong style render ([c928054](https://github.com/MarceloDJunior/react-optimized-images/commit/c928054290ff274bb8f29d34c67a57928d51edee))



## [1.2.1](https://github.com/MarceloDJunior/react-optimized-images/compare/v1.2.0...v1.2.1) (2022-05-29)


### Performance Improvements

* support only esm components to create a lighter bundle but keep script as cjs ([ca1dfa2](https://github.com/MarceloDJunior/react-optimized-images/commit/ca1dfa2be7803f4696e1d444bd452da4a741d2c9))



# [1.2.0](https://github.com/MarceloDJunior/react-optimized-images/compare/v1.1.1...v1.2.0) (2022-05-29)


### Features

* add different size images for better performance on mobile ([51fd933](https://github.com/MarceloDJunior/react-optimized-images/commit/51fd933cae1c7ec9523fedb047f41d406daef7ba))



## 1.1.1 (2022-05-25)


### Bug Fixes

* fix dependencies not being installed ([3506ed4](https://github.com/MarceloDJunior/react-optimized-images/commit/3506ed43d74e1f97d8887de46fba03dc0b2f0367))


### Features

* add Picture component ([7a184f5](https://github.com/MarceloDJunior/react-optimized-images/commit/7a184f52b1167c33d3d6199411c31f799fca0a56))
* add webp generator script ([e2b1295](https://github.com/MarceloDJunior/react-optimized-images/commit/e2b129530d6aa2ac903d756c9ff395f2e7804cd7))
* handle image loading errors ([39633ad](https://github.com/MarceloDJunior/react-optimized-images/commit/39633ad5624106eea9be8d1dd768e4055936568b))





plugin/webpack-plugin.js
const fs = require('fs');

const generateImages = require(`${__dirname}/generate-images.js`);

const NEW_CONFIG_OUTPUT = `${__dirname}/../config.json`;

const FONT_COLORS = {
  BLUE: '\x1b[36m',
  DEFAULT: '\x1b[0m',
  GREEN: '\x1b[32m',
};

const defaultConfig = {
  minWidth: 200,
  breakpoints: [
    {
      maxWidth: 576,
      resizeTo: 50,
    },
    {
      maxWidth: 992,
      resizeTo: 70,
    },
  ],
  enabled: true,
  lazy: false,
};

class OptimizedImagesPlugin {
  options = {};

  constructor(options) {
    const customOptions = {
      minWidth: options?.minWidth ?? defaultConfig.minWidth,
      breakpoints: options?.breakpoints ?? defaultConfig.breakpoints,
      enabled: options?.enabled ?? defaultConfig.enabled,
      lazy: options?.lazy ?? defaultConfig.lazy,
    };
    this.options = customOptions;
  }

  apply(compiler) {
    if (this.options.enabled) {
      compiler.hooks.environment.tap('OptimizedImagesPlugin', () => {
        this.applyCustomConfig();
      });

      compiler.hooks.assetEmitted.tap(
        'OptimizedImagesPlugin',
        (_, { targetPath }) => {
          const extension = this.getExtension(targetPath);
          if (['jpg', 'jpeg', 'png'].includes(extension)) {
            generateImages(targetPath, this.options).then(() => {
              const fileName = targetPath.substring(
                targetPath.lastIndexOf('/') + 1
              );
              console.log(
                // eslint-disable-next-line max-len
                `${FONT_COLORS.BLUE}image${FONT_COLORS.DEFAULT} - ${fileName} ${FONT_COLORS.GREEN}optimized ✔`
              );
            });
          }
        }
      );
    }
  }

  applyCustomConfig() {
    const newJsonConfig = JSON.stringify(this.options, null, 2);
    fs.writeFileSync(NEW_CONFIG_OUTPUT, newJsonConfig);
  }

  getExtension(filePath) {
    return filePath.substring(filePath.lastIndexOf('.') + 1).toLowerCase();
  }
}

module.exports = OptimizedImagesPlugin;
